doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+method needs to check if something like shutdown is currently occurring.
+Suspending threads during shutdown is pointless as the shutdown process will
+wake them up anyway to terminate them. If the checkpoint should be interrupted,
+the thread returns immediately; otherwise, it proceeds to checkpoint[^1].
+
+Next it triggers compilation of any methods that should be compiled proactively
+and waits for them to finish. Then, it calls
+`TR::CompilationInfo::suspendJITThreadsForCheckpoint` which:
+
+1. Suspends Compilation Threads
+2. Suspends the Sampler Thread
+3. Suspends the IProfiler Thread
+
+It then frees any SSL certificates. Finally it sets the Checkpoint Status to
+`TR_CheckpointStatus::READY_FOR_CHECKPOINT_RESTORE`, which indicates that the
+compiler is now ready for checkpoint. Before returning it ensures that it
+releases the Comp Monitor and reacquires VM Access.
+
+As will become evident shortly, all of the subtlety and complexity comes from
+the combination of the facts that  **i.** shutdown, jitdump, or both can occur
+at any point between the start and end of `prepareForCheckpoint`,  **ii.**
+coordinating the checkpointing thread and the Compiler Thread(s) to be suspended
+involves using a monitor different from the one used to coordinate said Compiler
+Thread(s).
+
+## Monitors
+
+**Compilation Monitor:** Referred to below as "Comp Monitor"; the monitor used
+by the compilation infrastructure to coordinate compilation threads, compilation
+entries, etc. (not to be confused with the Compilation Thread Monitor, see
+[Compilation Threads](./CompilationControl.md#compilation-threads) and
+ [Compilation Monitor](./CompilationControl.md#compilation-monitor) for more
+details).
+
+**Compilation Thread Monitor:** Referred to below as "CompThread Monitor"; the
+monitor used by a compilation thread to suspend itself for a checkpoint, or to
+wait for work.
+
+**CR Monitor:** The monitor used by the Checkpointing Thread to wait until the
+Compiler Thread(s) have finished suspending themselves.
+
+**Sampler Monitor:** The monitor used to coordinate the Sampler Thread.
+
+**IProfiler Monitor:** The monitor used to coordinate the IProfiler Thread.
+
+## Suspending Compilation Threads
+
+To maintain coherency, though not every monitor needs to always be acquired,
+the monitors must always be acquired in the following order:
+1. Comp Monitor
+2. CompThread Monitor
+3. CR Monitor
+
+### Normal Execution (steps 5 & 6 can be interchanged)
+The Checkpointing Thread at this point has the Comp Monitor in hand.
+1. The Checkpointing Thread signals the Comp Threads to suspend
+2. The Checkpointing Thread acquires the CR Monitor, releases the Comp Monitor,
+   and waits on the CR Monitor
+3. The Comp Threads acquire the Comp Monitor and the CompThread Monitor, and
+   change their state to `COMPTHREAD_SUSPENDED`
+4. The Comp Threads acquire the CR Monitor, call `notifyAll`, and release the CR
```

Comment:
Related to my previous comment: How could Comp Threads acquires the CR Monitor since in step 2, `The Checkpointing Thread acquires the CR Monitor` already

Reply:
The Checkpointing Thread does acquire the CR Monitor, but it calls `wait()`, which causes the thread to
1. Release the monitor
2. Sleep (or spin, whatever the underlying implementation is) until some other thread calls `notify()` or `notifyAll()`.